sources/osg/Polytope.js

```diff
@@ -0,0 +1,356 @@
+
+            this._planeList.length = 0;
+            this._planeList.push( Plane.create().set( 1.0, 0.0, 0.0, 1.0 ) ); // left plane.
+            this._planeList.push( Plane.create().set( -1.0, 0.0, 0.0, 1.0 ) ); // right plane.
+            this._planeList.push( Plane.create().set( 0.0, 1.0, 0.0, 1.0 ) ); // bottom plane.
+            this._planeList.push( Plane.create().set( 0.0, -1.0, 0.0, 1.0 ) ); // top plane.
+            if ( withNear ) this._planeList.push( Plane.create().set( 0.0, 0.0, 1.0, 1.0 ) ); // near plane
+            if ( withFar ) this._planeList.push( Plane.create().set( 0.0, 0.0, -1.0, 1.0 ) ); // far plane
+            this.setupMask();
+        },
+
+
+        /** Create a Polytope which is a equivalent to BoundingBox.*/
+        setToBoundingBox: function ( bb ) {
+            this._planeList.length = 0;
+            this._planeList.push( Plane.create().set( 1.0, 0.0, 0.0, -bb.getMin()[ 0 ] ) ); // left plane.
+            this._planeList.push( Plane.create().set( -1.0, 0.0, 0.0, bb.getMax()[ 0 ] ) ); // right plane.
+            this._planeList.push( Plane.create().set( 0.0, 1.0, 0.0, -bb.getMin()[ 1 ] ) ); // bottom plane.
+            this._planeList.push( Plane.create().set( 0.0, -1.0, 0.0, bb.getMax()[ 1 ] ) ); // top plane.
+            this._planeList.push( Plane.create().set( 0.0, 0.0, 1.0, -bb.getMin()[ 2 ] ) ); // near plane
+            this._planeList.push( Plane.create().set( 0.0, 0.0, -1.0, bb.getMax()[ 2 ] ) ); // far plane
+            this.setupMask();
+        },
+
+        setAndTransformProvidingInverse: function ( pt, matrix ) {
+            this._referenceVertexList = pt._referenceVertexList;
+            var resultMask = pt._maskStack[ this._maskStack.length - 1 ];
+            if ( resultMask === 0 ) {
+                this._maskStack[ this._maskStack.length - 1 ] = 0;
+                this._resultMask = 0;
+                this._planeList.length = 0;
+                return;
+            }
+            var selectorMask = 0x1;
+
+            var numActivePlanes = 0;
+            // count number of active planes.
+            var i;
+            for ( i = 0; i !== pt._planeList.length; ++i ) {
+                if ( resultMask & selectorMask ) ++numActivePlanes;
+                selectorMask <<= 1;
+            }
+
+            this._planeList.length = numActivePlanes;
+            this._resultMask = 0;
+            selectorMask = 0x1;
+            var index = 0;
+            for ( i = 0; i !== pt._planeList.length; ++i ) {
+                if ( resultMask & selectorMask ) {
+                    this._planeList[ index ] = pt._planeList[ i ];
+                    Plane.transformProvidingInverse( this._planeList[ index++ ], matrix );
+                    this._resultMask = ( this._resultMask << 1 ) | 1;
+                }
+                selectorMask <<= 1;
+            }
+
+            this._maskStack[ this._maskStack.length - 1 ] = this._resultMask;
```

Comment:
we need a general stack we could add a TODO and an issue to fix this at one time